File: browser/src/js/moneysocket/core/message/notification/provider.js
// Copyright (c) 2020 Jarret Dyrbye
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php


const BinUtl = require('../../../utl/bin.js').BinUtl;
const Uuid = require('../../../utl/uuid.js').Uuid;
const MoneysocketNotification = require(
    './notification.js').MoneysocketNotification;
let NOTIFICATION_SUBCLASSES = require(
    './notification.js').NOTIFICATION_SUBCLASSES;

class NotifyProvider extends MoneysocketNotification {
    constructor(provider_uuid, request_reference_uuid, payer, payee, msats) {
        super("NOTIFY_PROVIDER", request_reference_uuid);
        this.provider_uuid = provider_uuid;
        this.payer = payer
        this.payee = payee
        this.msats = msats
    }

    cryptLevel() {
        return "AES";
    }

    static castClass(msg_dict) {
        var c = new NotifyProvider(msg_dict['provider_uuid'],
                                   msg_dict['request_reference_id'],
                                   msg_dict['payer'], msg_dict['payee'],
                                   msg_dict['msats']);
        Object.keys(msg_dict).forEach(key => {
            c[key] = msg_dict[key];
        });
        return c;
    }

    static checkValidMsgDict(msg_dict) {
        if ( !('provider_uuid' in msg_dict)) {
            return "no provider_uuid included";
        }
        if (typeof msg_dict['provider_uuid'] != 'string') {
            return "unknown provider_uuid type";
        }
        if (! Uuid.isUuid(msg_dict['provider_uuid'])) {
            return "invalid provider_uuid";
        }
        if (typeof msg_dict['payee'] != "boolean") {
            return "payee must be True or False";
        }
        if (typeof msg_dict['payer'] != "boolean") {
            return "payer must be True or False";
        }
        if (msg_dict['msats'] != null) {
            if (typeof msg_dict['msats'] != 'number') {
                return "msats must be an integer";
            }
            if (msg_dict['msats'] < 0) {
                return "msats must be a positive value";
            }
        }
        return null;
    }
}

NOTIFICATION_SUBCLASSES['NOTIFY_PROVIDER'] = NotifyProvider;

exports.NotifyProvider = NotifyProvider;

File: python/moneysocket/core/message/notification/provider.py
# Copyright (c) 2020 Jarret Dyrbye
# Distributed under the MIT software license, see the accompanying
# file LICENSE or http://www.opensource.org/licenses/mit-license.php

import string
import json

from moneysocket.core.message.notification.notification import (
    MoneysocketNotification)

class NotifyProvider(MoneysocketNotification):
    MUST_BE_CLEARTEXT = False

    def __init__(self, provider_uuid, request_reference_uuid=None, payer=False,
                 payee=False, msats=None):
        super().__init__("NOTIFY_PROVIDER",
                         request_reference_uuid=request_reference_uuid)
        self['provider_uuid'] = str(provider_uuid)
        # will this provider pay outgoing invoices
        self['payer'] = payer
        # will this provider generate invoices for incoming payments
        self['payee'] = payee
        # balance to advertize as being available
        self['msats'] = msats

    @staticmethod
    def cast_class(msg_dict):
        c = NotifyProvider(msg_dict['provider_uuid'],
            request_reference_uuid=msg_dict['request_reference_uuid'],
            payer=msg_dict['payer'], payee=msg_dict['payee'],
            msats=msg_dict['msats'])
        c.update(msg_dict)
        return c

    @staticmethod
    def check_valid_msg_dict(msg_dict):
        if 'provider_uuid' not in msg_dict.keys():
            return "no provider_uuid included"
        if type(msg_dict['provider_uuid']) != str:
            return "unknown provider_uuid type"
        try:
            _ = uuid.UUID(msg_dict['provider_uuid'])
        except:
            return "invalid provider_uuid"

        if type(msg_dict['payee']) != bool:
            return "payee must be True or False"
        if type(msg_dict['payer']) != bool:
            return "payer must be True or False"
        if msg_dict['msats'] != None:
            if type(msg_dict['msats']) != int:
                return "msats must be an integer"
            if msg_dict['msats'] < 0:
                return "msats must be a positive value"
        return None


MoneysocketNotification.NOTIFICATION_SUBCLASSES['NOTIFY_PROVIDER'] = (
    NotifyProvider)
